examples/step-57/step-57.cc

```diff
@@ -0,0 +1,1101 @@
+    	{
+    		++outer_iteration;
+    		if (first_step)
+    		{
+    		  setup_system();
+    		  assemble_NavierStokes_system(first_step);
+    		  solve(first_step);
+    		  present_solution.add(1, newton_update);
+    		  nonzero_constraints.distribute(present_solution);
+    	      process_solution();
+    		  current_res = compute_residual(0);
+    		  first_step = false;
+    		  std::cout << "******************************" << std::endl;
+    		  std::cout << " The residual of initial guess is " << current_res << std::endl;
+    	      std::cout << " Initialization complete!  " << std::endl;
+       	    }
+
+    	  else
+          {
+    		assemble_NavierStokes_system(first_step);
+            last_res = system_rhs.l2_norm();
+            solve(first_step);
+            double alpha = 1.0;
+            // Because of the rough computation of update term, the Newton's iteration, $x_{new} = x_{old} + x_{update}$,
+            // probably fails to show us an improved solution at present step. Hence the weight of update term is moderated
+            // to let the present solution be better. Starting with 1, we take its half if the current residual is not
+            // better than the previous one until an appropriate weight is found. In this case, the Newton's iteration
+            // acts as $x_{new} = x_{old} + \alpha x_{update}$, where $\alpha$ is less than 1.
+
+            for (alpha = 1.0; alpha > 1e-5; alpha *= 0.5)
+              {
+                current_res = compute_residual(alpha);
+                std::cout << " alpha = " << std::setw(6) << alpha << std::setw(0)
+                          << " res = " << current_res << std::endl;
+                if (current_res < last_res)
+                  break;
+              }
+
+            {
+              last_res = current_res;
+              present_solution.add(alpha, newton_update);
+              nonzero_constraints.distribute(present_solution);
+              process_solution();
+              std::cout << " ----The " << outer_iteration << "th iteration. ---- " << std::endl;
+              std::cout << " ----Residual: " << current_res << std::endl;
+            }
+
+          }
+    	output_results (15*refinement+outer_iteration);
+      }
+    	refine_mesh();
+    	current_res =1;
+     }
+    }
+
+    // If the viscosity is smaller than 1/1000, we have to first search for an initial guess via "staircase". What we
+    // should notice is the search is always on the initial mesh, that is the $8 \times 8$ mesh in this program.
+    // After the searching part, we just do the same as we did when viscosity is larger than 1/1000: run Newton's iteration,
+    // refine the mesh, transfer solutions, and again.
+    else
+    {
+    	std::cout << "       Searching for initial guess ... " << std::endl;
+    	search_initial_guess(100.0);
+
+    	std::cout << "       Computing solution with target viscosity ..." <<std::endl;
```

Comment:
is there a difference between this code block and the one above? You are duplicating a lot of code for sure!